osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs

```diff
@@ -26,54 +25,7 @@ public override void Filter(FilterCriteria criteria)
         {
             base.Filter(criteria);
 
-            bool match =
-                criteria.Ruleset == null ||
-                Beatmap.RulesetID == criteria.Ruleset.ID ||
-                (Beatmap.RulesetID == 0 && criteria.Ruleset.ID > 0 && criteria.AllowConvertedBeatmaps);
-
-            if (Beatmap.BeatmapSet?.Equals(criteria.SelectedBeatmapSet) == true)
-            {
-                // only check ruleset equality or convertability for selected beatmap
-                Filtered.Value = !match;
-                return;
-            }
-
-            match &= !criteria.StarDifficulty.HasFilter || criteria.StarDifficulty.IsInRange(Beatmap.StarDifficulty);
-            match &= !criteria.ApproachRate.HasFilter || criteria.ApproachRate.IsInRange(Beatmap.BaseDifficulty.ApproachRate);
-            match &= !criteria.DrainRate.HasFilter || criteria.DrainRate.IsInRange(Beatmap.BaseDifficulty.DrainRate);
-            match &= !criteria.CircleSize.HasFilter || criteria.CircleSize.IsInRange(Beatmap.BaseDifficulty.CircleSize);
-            match &= !criteria.Length.HasFilter || criteria.Length.IsInRange(Beatmap.Length);
-            match &= !criteria.BPM.HasFilter || criteria.BPM.IsInRange(Beatmap.BPM);
-
-            match &= !criteria.BeatDivisor.HasFilter || criteria.BeatDivisor.IsInRange(Beatmap.BeatDivisor);
-            match &= !criteria.OnlineStatus.HasFilter || criteria.OnlineStatus.IsInRange(Beatmap.Status);
-
-            match &= !criteria.Creator.HasFilter || criteria.Creator.Matches(Beatmap.Metadata.AuthorString);
-            match &= !criteria.Artist.HasFilter || criteria.Artist.Matches(Beatmap.Metadata.Artist) ||
-                     criteria.Artist.Matches(Beatmap.Metadata.ArtistUnicode);
-
-            match &= !criteria.UserStarDifficulty.HasFilter || criteria.UserStarDifficulty.IsInRange(Beatmap.StarDifficulty);
-
-            if (match)
-            {
-                var terms = Beatmap.SearchableTerms;
-
-                foreach (var criteriaTerm in criteria.SearchTerms)
-                    match &= terms.Any(term => term.Contains(criteriaTerm, StringComparison.InvariantCultureIgnoreCase));
-
-                // if a match wasn't found via text matching of terms, do a second catch-all check matching against online IDs.
-                // this should be done after text matching so we can prioritise matching numbers in metadata.
-                if (!match && criteria.SearchNumber.HasValue)
-                {
-                    match = (Beatmap.OnlineBeatmapID == criteria.SearchNumber.Value) ||
-                            (Beatmap.BeatmapSet?.OnlineBeatmapSetID == criteria.SearchNumber.Value);
-                }
-            }
-
-            if (match)
-                match &= criteria.Collection?.Beatmaps.Contains(Beatmap) ?? true;
-
-            Filtered.Value = !match;
+            Filtered.Value = !criteria.Matches(Beatmap);
```

Comment:
Not super sure about using this virtual directly. It makes it possible that a careless implementor will kneecap song select filtering due to incorrectly overriding `Matches()` (by e.g. ignoring the base). Maybe the built-in song select stuff should be non-virtual, and the custom stuff should be on top of it - something like

```csharp
internal bool Matches(BeatmapInfo beatmap)
{
	bool match = true;
	// check built-in logic
	match &= MatchesCustomFilter(beatmap);
	return match;
}

public virtual bool MatchesCustomFilter(BeatmapInfo beatmap) => true;
```

Reply:
I definitely prefer that.